text/0073-entity-slicing.md

```diff
@@ -0,0 +1,562 @@
+# RFC: Entity Slicing using Data Tries
+
+## Related Issues
+
+- Reference Issues: [#592](https://github.com/cedar-policy/cedar/issues/592)
+- Implementation PR(s): [#1102](https://github.com/cedar-policy/cedar/pull/1102)
+
+## Timeline
+
+* Start Date: 2024-06-17
+* Date Entered FCP:
+* Date Accepted:
+* Date Landed:
+
+## Summary
+
+This RFC proposes a new API for finding an **entity slice**: a subset of the entity store needed for a particular request.
+Specifically, we propose a new function with signature `(Schema, Policies) -> EntityManifest`. There are more details on the `EntityManifest` later in this document, but the key idea is to store data paths **indexed by the request type**. The `Policies` must validate against the `Schema` in strict mode.
+
+The rest of this document follows the following format:
+
+* **Motivation**
+* **Example**- example policies show what data paths are needed for what request types.
+* **Example Usage**- we’ll show an example use-case for the entity manifest to put the design in context.
+* **Data Structure-** we’ll give details of the entity manifest data structure.
+* **Computing the Entity Manifest-** we’ll describe the static analysis used to generate the entity manifest.
+* **SimplifiedEntityLoader API**- we’ll describe a simple way to use the entity manifest.
+* **Drawbacks**- the entity manifest has several limitations to discuss.
+
+## Motivation
+
+Applications using Cedar need to provide entity data along with an authorization request. However, it is cost-prohibitive to provide all of an application’s data on every request, so users will want to provide a minimum amount, a *slice*. This proposal demonstrates a way to provide a **sound** slice: users provide a subset of the data, but get the same answer as if the full dataset had been provided.
+
+The design of entity slicing in this document is meant to achieve several goals:
+
+1. **Easy to use**- Leveraging entity slicing should be easy for Cedar users in a variety of settings
+2. **Performant**- Implementing the slice constraints improves performance without introducing too much overhead.
+3. **Flexible**- Entity slicing should work on a variety of data storage formats.
+
+
+
+## Example
+
+Here is a schema involving users and documents:
+
+```
+entity User in [User];
+
+entity Metadata = {
+   owner: User,
+   time: String,
+};
+
+entity Document = {
+  metadata: Metadata,
+  readers: Set<User>,
+};
+
+action Read, Edit appliesTo {
+  principal: [User],
+  resource: [Document],
+};
+```
+
+
+Here are a couple cedar policies conforming to the schema:
```

Comment:
Can we have an example based on a `User` that has an attribute `manager` pointing to another object of entity type `User` with a policy such as:

```
permit(
  principal,
  action == Action::"Read",
  resource
)
when
{
 principal.manager,businnesUnit = "1234"
};
```

This will force to handle another way relationship in between entity can be expressed.

Reply:
@jeffsec-aws, is this similar to the existing example with `resource.metadata.owner`? I don't want to overcomplicate the example.

Reply:
yes @oflatt  my bad. Discard.